File: app/src/main/java/com/example/bunnix/database/firebase/collections/ReviewCollection.kt
package com.example.bunnix.database.firebase.collections

import com.example.bunnix.database.config.FirebaseConfig
import com.example.bunnix.database.models.Review
import com.google.firebase.Timestamp
import com.google.firebase.firestore.Query
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

object ReviewCollection {

    private val collection = FirebaseConfig.firestore.collection(FirebaseConfig.Collections.REVIEWS)

    // ADD REVIEW
    suspend fun addReview(review: Review): Result<String> {
        return try {
            val docRef = collection.add(review).await()

            // Update vendor rating
            updateVendorRating(review.vendorId)

            Result.success(docRef.id)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // GET VENDOR REVIEWS (Real-time)
    fun getVendorReviews(vendorId: String): Flow<List<Review>> = callbackFlow {
        val listener = collection
            .whereEqualTo("vendorId", vendorId)
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                val reviews = snapshot?.toObjects(Review::class.java) ?: emptyList()
                trySend(reviews)
            }
        awaitClose { listener.remove() }
    }

    // GET CUSTOMER REVIEWS
    fun getCustomerReviews(customerId: String): Flow<List<Review>> = callbackFlow {
        val listener = collection
            .whereEqualTo("customerId", customerId)
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                val reviews = snapshot?.toObjects(Review::class.java) ?: emptyList()
                trySend(reviews)
            }
        awaitClose { listener.remove() }
    }

    // ADD VENDOR RESPONSE
    suspend fun addVendorResponse(reviewId: String, response: String): Result<Unit> {
        return try {
            val updates = mapOf(
                "vendorResponse" to response,
                "vendorResponseAt" to Timestamp.now()
            )
            collection.document(reviewId).update(updates).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // UPDATE VENDOR RATING
    private suspend fun updateVendorRating(vendorId: String) {
        try {
            val reviews = collection
                .whereEqualTo("vendorId", vendorId)
                .get()
                .await()
                .toObjects(Review::class.java)

            if (reviews.isEmpty()) return

            val totalRating = reviews.sumOf { it.rating }
            val averageRating = totalRating.toDouble() / reviews.size

            // Round to 1 decimal place
            val roundedRating = String.format("%.1f", averageRating).toDouble()

            val vendorUpdates = mapOf(
                "rating" to roundedRating,
                "totalReviews" to reviews.size
            )

            FirebaseConfig.firestore
                .collection(FirebaseConfig.Collections.VENDOR_PROFILES)
                .document(vendorId)
                .update(vendorUpdates)
                .await()
        } catch (e: Exception) {
            // Ignore errors
        }
    }
}

File: src/database/firebase/collections/reviewCollection.js
import { db, COLLECTIONS } from '../config';
import { 
  collection, 
  doc,
  addDoc, 
  getDoc,
  getDocs,
  updateDoc,
  query, 
  where, 
  orderBy,
  onSnapshot,
  increment,
  serverTimestamp 
} from 'firebase/firestore';

const reviewsCollection = collection(db, COLLECTIONS.REVIEWS);

/**
 * ADD REVIEW
 */
export const addReview = async (reviewData) => {
  try {
    const docRef = await addDoc(reviewsCollection, {
      ...reviewData,
      createdAt: serverTimestamp(),
      updatedAt: serverTimestamp()
    });
    
    // Update vendor rating
    await updateVendorRating(reviewData.vendorId);
    
    return { success: true, reviewId: docRef.id };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * GET VENDOR REVIEWS (Real-time)
 */
export const subscribeToVendorReviews = (vendorId, callback) => {
  const q = query(
    reviewsCollection,
    where('vendorId', '==', vendorId),
    orderBy('createdAt', 'desc')
  );
  
  return onSnapshot(q, (snapshot) => {
    const reviews = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data()
    }));
    callback(reviews);
  });
};

/**
 * GET CUSTOMER REVIEWS (Real-time)
 */
export const subscribeToCustomerReviews = (customerId, callback) => {
  const q = query(
    reviewsCollection,
    where('customerId', '==', customerId),
    orderBy('createdAt', 'desc')
  );
  
  return onSnapshot(q, (snapshot) => {
    const reviews = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data()
    }));
    callback(reviews);
  });
};

/**
 * ADD VENDOR RESPONSE
 */
export const addVendorResponse = async (reviewId, response) => {
  try {
    await updateDoc(doc(reviewsCollection, reviewId), {
      vendorResponse: response,
      vendorResponseAt: serverTimestamp(),
      updatedAt: serverTimestamp()
    });
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * INCREMENT HELPFUL COUNT
 */
export const incrementHelpful = async (reviewId) => {
  try {
    await updateDoc(doc(reviewsCollection, reviewId), {
      helpful: increment(1)
    });
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * UPDATE VENDOR RATING
 */
const updateVendorRating = async (vendorId) => {
  try {
    const q = query(reviewsCollection, where('vendorId', '==', vendorId));
    const snapshot = await getDocs(q);
    
    if (snapshot.empty) return;
    
    const reviews = snapshot.docs.map(doc => doc.data());
    const totalRating = reviews.reduce((sum, review) => sum + review.rating, 0);
    const averageRating = totalRating / reviews.length;
    const roundedRating = Math.round(averageRating * 10) / 10; // Round to 1 decimal
    
    // Update vendor profile (import from vendorProfileCollection)
    const vendorProfilesCollection = collection(db, COLLECTIONS.VENDOR_PROFILES);
    await updateDoc(doc(vendorProfilesCollection, vendorId), {
      rating: roundedRating,
      totalReviews: reviews.length
    });
  } catch (error) {
    console.error('Error updating vendor rating:', error);
  }
};